src/project/internal/recentfilescontroller.cpp

```diff
@@ -0,0 +1,298 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "recentfilescontroller.h"
+
+#include <QtConcurrent>
+
+#include <QJsonArray>
+#include <QJsonDocument>
+
+#include "async/async.h"
+#include "defer.h"
+
+#include "multiinstances/resourcelockguard.h"
+
+using namespace mu::project;
+using namespace mu::async;
+
+static const std::string RECENT_FILES_RESOURCE_NAME("RECENT_FILES");
+
+void RecentFilesController::init()
+{
+    TRACEFUNC;
+
+    m_dirty = true;
+
+    multiInstancesProvider()->resourceChanged().onReceive(this, [this](const std::string& resourceName) {
+        if (resourceName == RECENT_FILES_RESOURCE_NAME) {
+            if (!m_reloadingBlocked) {
+                m_dirty = true;
+            }
+
+            m_recentFilesListChanged.notify();
+        }
+    });
+}
+
+const ProjectFilesList& RecentFilesController::recentFilesList() const
+{
+    TRACEFUNC;
+
+    if (m_dirty) {
+        const_cast<RecentFilesController*>(this)->loadRecentFilesList();
+    }
+
+    const_cast<RecentFilesController*>(this)->removeNonexistentFiles();
+
+    return m_recentFilesList;
+}
+
+Notification RecentFilesController::recentFilesListChanged() const
+{
+    return m_recentFilesListChanged;
+}
+
+void RecentFilesController::prependRecentFile(const ProjectFile& newFile)
+{
+    if (!newFile.isValid()) {
+        return;
+    }
+
+    TRACEFUNC;
+
+    ProjectFilesList newList;
+    newList.reserve(m_recentFilesList.size() + 1);
+    newList.push_back(newFile);
+
+    for (const ProjectFile& file : m_recentFilesList) {
+        if (file.path != newFile.path && fileSystem()->exists(file.path)) {
+            newList.push_back(file);
+        }
+    }
+
+    setRecentFilesList(newList, true);
+
+    prependPlatformRecentFile(newFile.path);
+}
+
+void RecentFilesController::moveRecentFile(const io::path_t& before, const ProjectFile& after)
```

Comment:
Why are these arguments of a different type?

Reply:
Only the file path is really a key, and the displayNameOverride is some additional data. So, to look up the existing file in the list, we should only compare the path.